Theme tree list

jmuzina opened the pull request:
## Done

[List of work items including drive-bys]

Fixes #5269, [WD-11827](https://warthogs.atlassian.net/jira/software/c/projects/WD/boards/970?assignee=712020%3A8af91202-4b48-4d2b-997f-165f9c76300e&selectedIssue=WD-11877), [WD-13882](https://warthogs.atlassian.net/browse/WD-13882)

## QA

- Open [demo](insert-demo-url)
- [Add QA steps]
- Review updated documentation:
  - [List any updated documentation for review]

### Check if PR is ready for release

If this PR contains Vanilla SCSS code changes, it should contain the following changes to make sure it's ready for the release:

- [x] PR should have one of the following labels to automatically categorise it in release notes:
  - `Feature 🎁`, `Breaking Change 💣`, `Bug 🐛`, `Documentation 📝`, `Maintenance 🔨`.
- [ ] Vanilla version in `package.json` should be updated relative to the [most recent release](https://github.com/canonical/vanilla-framework/releases/latest), following semver convention:
  - if CSS class names are not changed it can be bugfix release (x.x.**X**)
  - if CSS class names are changed/added/removed it should be minor version (x.**X**.0)
  - see the [wiki for more details](https://github.com/canonical/vanilla-framework/wiki/Release-process#pre-release-tasks)
- [ ] Any changes to component class names (new patterns, variants, removed or added features) should be listed on the [what's new page](https://github.com/canonical/vanilla-framework/blob/main/releases.yml).


## Screenshots

[if relevant, include a screenshot or screen capture]

webteam-app commented:
### [<img src='https://assets.ubuntu.com/v1/6baef514-ubuntu-circle-of-friends-large.svg' height=32 width=32> Demo</img>](https://vanilla-framework-5270.demos.haus)
### [<img src='https://assets.ubuntu.com/v1/e512b0e2-jenkins.svg' height=32 width=32> Jenkins </img>](https://jenkins.canonical.com/webteam/job/start-demo/8751)
### [<img src='https://assets.ubuntu.com/v1/7144ec6d-logo-jaas-icon.svg' height=32 width=32> demos.haus </img>](https://demos.haus)

lyubomir-popov commented:
looks good, just a couple of notes:
- the chevron's bounding box should align with the preceding items, the way this red chevron does - so increase the indent a little (the sidenav already does this I think)
 
![image](https://github.com/user-attachments/assets/0e2c9422-8956-4da7-aca3-3ba2d04b84b9)


active should appear just once, on the leaf node corresponding to the page view displayed in a hypothetical adjacent panel controlled by this componoent. I suggest bold as the simplest thing, that doesn't get in theway of indents

jmuzina commented:
> active should appear just once, on the leaf node corresponding to the page view displayed in a hypothetical adjacent panel controlled by this component.

@lyubomir-popov I'm not certain about this - doesn't this change this component to function more like side navigation, rather than as a list? Maybe we're starting to muddy this component's purpose. My best UX understanding here is that the active state is there to help the user trace through the open tree structure, not to indicate a specific active node. 

I'm happy to implement it, just want to be sure I have a clear understanding of this component's purpose.

lyubomir-popov commented:
spec should answer https://www.w3.org/WAI/ARIA/apg/patterns/treeview/

jmuzina commented:
See summary of meeting with @akbarkz on sites' use case [here](https://warthogs.atlassian.net/browse/WD-11877?focusedCommentId=547306).

jmuzina commented:
@lyubomir-popov Updated - have a look again please! Be sure to check on small and large screens (indentation changes to align with side navigation)

bartaz commented on scss/_patterns_list-tree.scss:
```diff
@@ -46,26 +42,32 @@
-    padding-left: 0.8rem;
+    padding-left: map-get($grid-margin-widths, small);
+
+    @media (min-width: $threshold-4-6-col) {
```
In side navigation this may have something to do with grid alignment (the grid margin is responsive - smaller on mobile), and the indentation of the side nav may align with this (or at least this may have been intention).

Does it need to apply to all nesting levels (or only top layer that may need alignment with the grid) I don't know.

I guess for the sake of updating this component we can attempt to implement simpler version first, and either wait for design review from Lyubo, or update it later when he can give some design input on it.

jmuzina replied:
Removed the responsive layout changes

jmuzina commented on scss/_patterns_list-tree.scss:
```diff
@@ -2,27 +2,26 @@
     list-style-type: none;
-    margin-left: $sp-medium;
-    padding: 0 0 0 $sp-x-small;
+    margin-left: $sp-x-small;
```
Removed the padding here

jmuzina commented:
> High-level comment: there's a minor metrics thing going on here where the left padding on list tree items isn't sufficient to actually separate the chevron area from the text; this could probably be smoothed out to make the metrics of everything more predictable:

@pastelcyborg I've adjusted the margin slightly to eliminate this issue:
![image](https://github.com/user-attachments/assets/a4fca9bb-4bd1-4a88-8681-1917c276bd9a)

jmuzina commented on scss/_patterns_list-tree.scss:
```diff
@@ -46,26 +42,32 @@
 
     margin-top: $sp-xx-small;
-    padding-left: 0.8rem;
+    padding-left: map-get($grid-margin-widths, small);
```
I managed to unify this by moving all the padding to the items themselves, rather than the parent. It's important that the items themselves have padding so that the items with toggle icons line up horizontally with the ones that don't.

jmuzina commented:
> I still see multiple parents with active states:
> 
> ![image](https://github.com/user-attachments/assets/3d271937-df2a-4f9c-827f-a0176e327ba8)

@lyubomir-popov 
Acknowledged, it is currently set to highlight ancestor nodes rather than just the parent of the active node.

Separately, links and groupings have an active state that color them blue. 

Would you like just the parent node to be highlighted grey, or none of the nodes to be highlighted grey?

lyubomir-popov commented:
Does the wcag spec require highlighting parents? If not, they are simply distracting.

jmuzina commented:
> Does the [wcag tree view spec](https://www.w3.org/WAI/ARIA/apg/patterns/treeview/) require highlighting parents? If not, they are simply distracting.

@lyubomir-popov It does not - removed the highlighting

jmuzina commented on templates/docs/examples/patterns/list-tree.html:
```diff
@@ -56,14 +52,58 @@
+ * @param {HTMLElement} listTreeLink - `.p-list-tree__toggle` or `.p-list-tree__item` element to set up.
+ * @param {HTMLElement[]} otherTreeLinks - Other `.p-list-tree__toggle` or `.p-list-tree__item` elements on the page.
+ */
+function setupListTreeLink(listTreeLink, otherTreeLinks) {
```
This behavior is intentional, to make the component behave a little bit more like a navigational component. I used [Carbon](https://react.carbondesignsystem.com/?path=/story/components-treeview--default) and [MUI](https://mui.com/x/react-tree-view/) as examples here.

As for the `is-active` styling vs using `aria-expanded`, I'm not sure if this can be achieved without `is-active`. This component was changed so that one of the nodes can be the "active" node (controlling some primary content to its right). The links themselves are not currently highlighted as active, only the grouping nodes (which have the `aria-expanded`).

jmuzina replied:
@pastelcyborg That's correct - my language was a bit imprecise, sorry. When I said "currently" I was referring to the current version of this component as released. 

In the released version, the groupings are highlighted grey when expanded, and there is no active styling for the nodes, just default link styling if the nodes are `a` elements.
<img width="232" alt="Screenshot 2024-08-29 at 3 10 47 PM" src="https://github.com/user-attachments/assets/c4a1dd76-b28a-4f81-9a79-926b2f166934">

In this proposed version, the groupings are never highlighted grey, and the nodes are colored as default text, unless they are active.

<img width="232" alt="Screenshot 2024-08-29 at 3 12 26 PM" src="https://github.com/user-attachments/assets/e40a692c-b6b1-41d3-8cf5-502ac179774d">

jmuzina commented on scss/_patterns_list-tree.scss:
```diff
@@ -46,26 +56,20 @@
-      top: 0.8rem;
-      width: 0.625rem;
+  .p-list-tree__toggle[aria-expanded='false'] {
+    &::after {
```
Changed to `before`

jmuzina commented on scss/_patterns_list-tree.scss:
```diff
@@ -2,42 +2,52 @@
+    cursor: pointer;
+
+    &:hover {
+      background: transparent;
```
Removed

jmuzina commented:
Notes from SU design review:

- Need more space between icon and text - see side nav
- Remove grey click color from the toggle buttons (inherited from button element)

jmuzina commented:
Removed the grey coloring - need to adjust the spacing as well, will do that early Monday

jmuzina commented:
> I see there was a note about increasing the spacing between the chevron and text, but this is now a huge amount of space that looks quite unusual, especially when this component is used as a file tree:
> That seems like way too much space to me and makes this component less useful for file tree use cases.

@lyubomir-popov Asked me to align this with side-navigation - this is roughly the same space between text and icon as sidenav (24px/1.5rem).

@lyubomir-popov Have a look and let me know what you think please

lyubomir-popov commented:
> Have a look and let me know what you think please

the spacing is actually 1rem/16px, both in the sidenav and the accordion:
<img width="452" alt="image" src="https://github.com/user-attachments/assets/5e854ea4-06e9-4b3a-a83e-05feaff32e3e">

reducing to that value eshould keep things consistent across component and pleasinfg to the eye IMO

jmuzina commented:
> > Have a look and let me know what you think please
> 
> the spacing is actually 1rem/16px, both in the sidenav and the accordion: <img alt="image" width="452" src="https://private-user-images.githubusercontent.com/2741678/362379497-5e854ea4-06e9-4b3a-a83e-05feaff32e3e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ4NjQwMzgsIm5iZiI6MTcyNDg2MzczOCwicGF0aCI6Ii8yNzQxNjc4LzM2MjM3OTQ5Ny01ZTg1NGVhNC0wNmU5LTRiM2EtYTgzZS0wNWZlYWZmMzJlM2UucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgyOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MjhUMTY0ODU4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NDE3NzU0MTMxZDcwNjBlNzBmNjk0ZjY1Yjk3ZTFkMGZmZjZjNDUyZTlmMTZkNGM2ZWQ1YjQzNGJhYzgyZDBjYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.mUWggyGWzBYmawaixbCvZLkdxHDv-jp-De-6EPutAV0">
> 
> reducing to that value should keep things consistent across component and pleasinfg to the eye IMO

Ah, must've made a mistake with my ruler - will change it to 1rem, good catch

jmuzina commented:
@lyubomir-popov Revised to `1.5em`/`16px`: 
<img width="918" alt="Screenshot 2024-08-28 at 1 06 56 PM" src="https://github.com/user-attachments/assets/fdbb4eb2-2e89-4cee-8420-18f13110afeb">

jmuzina commented:
@pastelcyborg Per our quick meeting just now, I revised to move the links into link elements inside the list items. Addressing your points below to make sure everything is covered:

>     1. Maybe there should be an equivalent child element to `p-list-tree__toggle` for non-toggle elements that goes inside the `p-list-tree__item` and applies some form of inline styling?

I've added `.p-list-tree__link`, a child of `.p-list-tree__item`, that performs this function.

>     2. Related to 1, the additional left-padding on `p-list-tree__item` makes the click target have strange metrics

I think you mentioned that what I did above fixes this issue, but let me know if not.

>     3. Also related to 1, if the `p-list-tree__item` had a proper child element, it would not only solve the metrics issues, it would also be properly keyboard-navigable - right now, you can't tab through the child items

Good call, making these links inside the list elements does make the entire tree much more keyboard navigable.

bartaz commented on templates/docs/examples/patterns/list-tree.html:
```diff
@@ -56,14 +74,58 @@
+ * @param {String} className - The class to add or remove.
+ * @param {Boolean} add - Whether to add or remove the class.
+ */
+function setClass(element, className, add) {
```
Why are we adding an abstraction over a single line of code? Can't we use `element.classList.add/remove/toggle` directly?

jmuzina replied:
Removed the abstraction

bartaz commented:
I've seen mentions of some potential class name and markup changes. Are we doing those? 

We need to be careful not to introduce breaking changes in such case. The places that currently use the component will get a new styling, but not new HTML/JS. If any markup changes are introduced, we need to keep a "legacy" example with previous markup to make sure we leave it in "usable" state.

jmuzina commented:
> I've seen mentions of some potential class name and markup changes. Are we doing those?

We have changed the leaf nodes so that they always contain an interactive link element, with `.p-list-tree__link`. JS was modified to add the requested active link highlighting.

> We need to be careful not to introduce breaking changes in such case. The places that currently use the component will get a new styling, but not new HTML/JS. If any markup changes are introduced, we need to keep a "legacy" example with previous markup to make sure we leave it in "usable" state.

Functionality of the [legacy](http://0.0.0.0:8101/docs/examples/patterns/list-tree/legacy?theme=light) version (old HTML/JS, new CSS) appears to be maintained, and there is a description of the differences between the versions [here](https://vanilla-framework-5270.demos.haus/docs/patterns/list-tree#legacy-version). Is that ok?

bartaz commented:
MAAS UI uses the tree with some custom style overrides:

https://github.com/canonical/maas-ui/blob/ee87049981cfc9edce97539deb58bccc729e6791/src/app/controllers/views/ControllerDetails/ControllerSummary/ServicesCard/ServicesCard.tsx#L109

https://github.com/canonical/maas-ui/blob/ee87049981cfc9edce97539deb58bccc729e6791/src/scss/_patterns_list-trees.scss#L4-L9

We should reach out (at least open an issue) that the component is changing to make sure we wont break too much in their use case.

jmuzina commented on scss/_patterns_list-tree.scss:
```diff
@@ -1,43 +1,48 @@
+    width: $list-tree-toggle-icon-size;
+  }
+
+  .p-list-tree__item:not(.p-list-tree__item--group) > .p-list-tree__link,
```
Cleaned this up.

jmuzina commented on scss/_patterns_list-tree.scss:
```diff
@@ -46,26 +51,21 @@
-  .p-list-tree__item--group::after {
-    @extend %list-tree-icon;
+  // Highlight active nodes
+  .p-list-tree__item.is-active:not(.p-list-tree__item--group) > .p-list-tree__link,
```
Cleaned this up.

jmuzina commented:
> MAAS UI uses the tree with some custom style overrides

Here's the before-after of the legacy list-tree markup/JS with the additional SCSS from MAAS:

Before:
<img width="227" alt="Screenshot 2024-09-04 at 3 02 54 PM" src="https://github.com/user-attachments/assets/537c1ea6-de5d-4882-ab90-9986e7d0649e">
Their styling forces the tree to be fully expanded, non-interactive, and to hide the collapse button.

After:
<img width="227" alt="Screenshot 2024-09-04 at 3 02 58 PM" src="https://github.com/user-attachments/assets/72597728-4d74-475d-8d7d-432f094f204b">
Keeping the same markup/JS & using our new SCSS, the tree's icons are changed to chevrons. The full tree is still expanded, but the "collapse" chevrons are now visible on toggle, which probably isn't expected. This is because their custom styling is hiding the minus icon, which was  `.p-list-tree::after` but is now  `.p-list-tree::before` for semantic reasons.

They should have identical behavior (but different styling) by adjusting their style to hide the collapse button so that it hides our collapse button, and replaces it with the expand button.

```scss
.p-list-tree--static {
    display: block !important;

    .p-list-tree--static::before {
      display: none !important;
    }

    .p-list-tree--static::after {
      @extend %list-tree-icon;
      @include vf-icon-chevron-right-themed;
    }
  }
```

I'll write this up as a separate issue on their board/repo.

bartaz reviewed (approved):
Looks good, thanks!

bartaz commented:
There may have been some hiccup with Percy, as it seems it didn't run?

jmuzina commented:
> There may have been some hiccup with Percy, as it seems it didn't run?

It was intentionally skipped to save Percy screenshots as I foresaw more changes coming - it's been run and the results are now [available](https://percy.io/bb49709b/vanilla-framework/builds/36252633?utm_campaign=bb49709b&utm_content=vanilla-framework&utm_source=github_status_private)

jmuzina commented:
Hi @akbarkz - this is nearing merge, have a look at [dark list tree](https://vanilla-framework-5270.demos.haus/docs/examples/patterns/list-tree/default?theme=dark) and [docs](https://vanilla-framework-5270.demos.haus/docs/patterns/list-tree) and let me know if you have any thoughts / concerns :)
